### ISSUE TITLE
StopWhenUnneeded + umounting busy filesystem leads to inconsistent state

### ISSUE THREAD

**Issue by aroig** [systemd/systemd#4146]
### Submission type
- [x] Bug report
- [ ] Request for enhancement (RFE)

_NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!_
### systemd version the issue has been seen with

> 231

_NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!_
### Used distribution

> arch linux
### In case of bug report: Expected behaviour you didn't see

> Assume a unit `mnt.mount` requires `mnt.service`, which sets StopWhenUnneeded=true. If `mnt.mount` is stopped while the filesystem is busy, both `mnt.mount` and `mnt.service` should remain active.
### In case of bug report: Unexpected behaviour you saw

> When attempting to stop an active `mnt.mount` while the filesystem is busy, the operation fails. `mnt.mount` remains active, but `mnt.service` is stopped, breaking the Requires dependency.
### In case of bug report: Steps to reproduce the problem
1. Prepare the following units.
   mnt.mount
   
   ```
   [Unit]
   Requires=mnt.service
   After=mnt.service
   
   [Mount]
   What=/dev/sdb1
   Where=/mnt
   ```
   
   and mnt.service
   
   ```
   [Unit]
   StopWhenUnneeded=true
   
   [Service]
   Type=oneshot
   RemainAfterExit=true
   
   ExecStart=/usr/bin/true
   ```
2. Then do the following
   
   ```
   $ systemctl start mnt.mount
   $ cd /mnt                  # to keep the filesystem busy
   $ systemctl stop mnt.mount
   $ systemctl status mnt.mount mnt.service
   ```

**Comment by poettering** [comment]
Hmm, in this case it appears you should be using BindsTo= instead of Requires=, no?

**Comment by aroig** [comment]
> Hmm, in this case it appears you should be using BindsTo= instead of Requires=, no?

No, the mount depends on the oneshot service, with `RemainAfterExit=yes`. I don't think the `BindsTo` would make any difference. The service would still be stopped, and the busy mount would not. Unless I'm missing something.

The actual problem is that a mount that refuses to umount, still propagates a stop on the required service, due to the `StopWhenUnneeded=true`. But then the mount remains active because it is busy, while the required dependency is brought down.

Propagating a stop due to `StopWhenUnneeded` should make sure that the stop on the depending units have succeeded. I guess mounts are special, in the sense that a failed stop leaves them active.

**Comment by poettering** [comment]
I see.

But I am not convinced StopWhenUnneeded= should really do what you are saying. I mean, the only thing it's really supposed to cover is when a unit is not needed by any other unit anymore. It's not supposed to be influenced by other ways how busy/non-busy might be communicated...

It appears to me that a construct involving an automount unit would be more appropriate here...

**Comment by aroig** [comment]
hmm... I do not follow what you mean.

>  I mean, the only thing it's really supposed to cover is when a unit is not needed by any other unit anymore

If the mount unit having `Requires=mnt.service` is active, implies that `mnt.service` is still needed. Then a single job is queued, which ends up stopping `mnt.service` while `mnt.mount` remains active. It looks to me that this already contradicts the quoted line above, isn't it? `mnt.service` does not die or anything, it is systemd that stops it mistakenly assuming it is not needed, while it still is.

> It appears to me that a construct involving an automount unit would be more appropriate here...

That would not work for my use case. I have an automount for `~/priv`, which mounts a luks backed encrypted partition. the associated mount unit requires a service to setup the luks things, unencrypt a passphrase by querying my user's gpg-agent and unlock the luks device (so, I can't use crypttab for this).

The good thing about it, is that `~/priv` unmounts and locks itself back after a short timeout. The bad thing is that if I try to umount a busy mountpoint, the luks service is brought down while the mount is still active, and things break. They do not break badly, the luks unit gets out of sync with kernel state.

Anyway, disregarding my particular use, the thing is that mount units and `StopWhenUnneeded` do not play well, and break the expectations of `Requires` and `BindsTo`. From a user perspective, this is weird. If you do not intend to fix it (I understand a fix might not be trivial), consider documenting it somehow. 

EDIT: I understood what you meant with the automount.

**Comment by aroig** [comment]
Ok, now that I think about this again, Maybe this has nothing to do with StopWhenUnneeded after all...

In the problem at hand, stopping the mount propagates a stop signal for the Required dependency. That's fine.

If the mount has, in addition, `After=mnt.service` that means that the stop signal for the required service must wait for the mount's stop process to finish. Of course, for a mount unit, when the stop process fails, the unit remains active.

Then, this issue could be fixed as follows: If we are waiting for a stop signal on a unit, check that the unit is truly stopped after the stop process finishes, and cancel the transaction if it is not. After all, systemd already does that for the start signals in a Requires/After pair.

This will keep the semantics of Requires and Afters (mostly) orthogonal, which maybe was the source of your original doubts...
